### PR TITLE
UCM/HOOKS: added ucm_test_events routine to check if hooks works properly

### PR DESCRIPTION
**Pull request by hoopoepg**
- added ucm_test_events call
- added gtest test for new call

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9189/ for details (Mellanox internal link).

File: src/ucm/mmap/install.c [hunk 1]
```diff
@@ -142,6 +132,24 @@ static ucs_status_t ucm_mmap_test(int events)
...
+    ucs_status_t status;
+
+    if (ucs_test_all_flags(installed_events, events)) {
+        /* All requested events are already installed */
+        return UCS_OK;
+    }
+
+    status = ucm_test_events(events);
+    if (status == UCS_OK) {
+        installed_events |= events;
```
Comment (brminich) [inline src/ucm/mmap/install.c]: We do not install events partially now. Is it ok?

Comment (hoopoepg) [reply src/ucm/mmap/install.c]: yes, this call doesn't install hooks at all - just trying catch events on system API call

Comment (brminich) [reply src/ucm/mmap/install.c]: yes, but you will not cache added events, like it was done before. So the next call with less events will go thru the whole check as well

Comment (hoopoepg) [reply src/ucm/mmap/install.c]: right, but this way simplifier logic :)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9195/ for details (Mellanox internal link).

File: src/ucm/mmap/install.c [hunk 2]
```diff
@@ -102,18 +102,11 @@ void ucm_fire_mmap_events(int events)
     }
 }
 
-/* Called with lock held */
-static ucs_status_t ucm_mmap_test(int events)
+ucs_status_t ucm_test_events(int events)
```
Comment (yosefe) [inline src/ucm/mmap/install.c]: don't we need to test also malloc/free?

File: src/ucm/mmap/install.c [hunk 3]
```diff
@@ -142,6 +132,24 @@ static ucs_status_t ucm_mmap_test(int events)
     return UCS_OK;
 }
 
+/* Called with lock held */
+static ucs_status_t ucm_mmap_test(int events)
```
Comment (yosefe) [inline src/ucm/mmap/install.c]: the API function should call internal helper, not the other way around..

**Comment by hoopoepg** [comment]
@brminich anything else?

File: src/ucm/mmap/install.c (hunk 2 above)
Comment (hoopoepg) [reply src/ucm/mmap/install.c]: don't think so - malloc/free are used for internal implementation, not an event

File: src/ucm/mmap/install.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucm/mmap/install.c]: fixed

File: src/ucm/mmap/install.c (hunk 1 above)
Comment (hoopoepg) [reply src/ucm/mmap/install.c]: restored original algo

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9220/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9223/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe more comments?

File: src/ucm/mmap/install.c [hunk 4]
```diff
@@ -102,46 +102,60 @@ void ucm_fire_mmap_events(int events)
...
     }
 
     return UCS_OK;
 }
 
+ucs_status_t ucm_test_events(int events)
+{
+    int out_events;
+
+    return ucm_mmap_test_events(events, &out_events);
```
Comment (yosefe) [inline src/ucm/mmap/install.c]: ucm_mmap_test -> `/* Called with lock held */`
ucm_mmap_test_events is called from ucm_mmap_test -> also need such comment ?
this function is API, should it take lock?

also, should't we test malloc events as well?

File: src/ucm/mmap/install.c [hunk 5]
```diff
@@ -102,46 +102,60 @@ void ucm_fire_mmap_events(int events)
...
     /* Return success iff we caught all wanted events */
-    if (!ucs_test_all_flags(out_events, events)) {
+    if (!ucs_test_all_flags(*out_events, events)) {
         return UCS_ERR_UNSUPPORTED;
     }
 
     return UCS_OK;
 }
 
+ucs_status_t ucm_test_events(int events)
```
Comment (yosefe) [inline src/ucm/mmap/install.c]: this function should be in event.c

Comment (hoopoepg) [reply src/ucm/mmap/install.c]: shifted

File: src/ucm/mmap/install.c (hunk 4 above)
Comment (hoopoepg) [reply src/ucm/mmap/install.c]: > ucm_mmap_test_events is called from ucm_mmap_test -> also need such comment ?

no, it should work out of lock - we are reproducing actual case which may be caused by user

there are no malloc events

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9243/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9267/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe anything else to fix?

**Pull request merged**